### ISSUE TITLE
Upgrade broke websockets

### ISSUE THREAD

**Issue by ChrisFernandez** [mattermost/platform#6168]
https://forum.mattermost.org/t/upgraded-from-3-7-to-3-8-broke-websockets-help/3163
Hello everything was working fine.. I upgraded like I always do... log in.. got a msg to add a URL in the console/configuration field.. I added my url... and after that I get:
websocket connect err: websocket: origin not allowed
/api/v3/users/websocket:connect code=500 rid=4che7w3383r88qxswssi53nkwh uid= ip=107.182.x.x Failed to upgrade websocket connection [details: ]

I changed the URL to all my interfaces and restarted with same issue...

MY setup?

nginx port 443 port 80 redirect to 443
config for nginx:

```
server {
      listen 443 ssl;
      server_name chat.mydomain.info o5imyonion.onion chat.mydomain;

      ssl on;
      ssl_certificate  /etc/letsencrypt/live/chat.mydomain.info/fullchain.pem;
      ssl_certificate_key /etc/letsencrypt/live/chat.mydomain.info/privkey.pem;
      ssl_dhparam /home/mattermost/cert/dhparam.pem;
      ssl_session_timeout 5m;
      ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
      ssl_ciphers 'EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH';
      ssl_prefer_server_ciphers on;
      ssl_session_cache shared:SSL:10m;


      location / {
        gzip off;
        proxy_set_header X-Forwarded-Ssl on;
        client_max_body_size 50M;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Frame-Options SAMEORIGIN;
        proxy_pass http://127.0.0.1:8065;
     }
   }

```
on mattermost I have added this two changes..
on the URL "enforced" field now..(I dont know why you guys force this.. when it SHOULD be an option to admins... not you...
I have http://127.0.0.1 (had my external ip, my external domains all of them, this is just the last one I tried)
for port I have:
127.0.0.1:8065

why? nginx is forwarding the connection to localhost:8065 as you can see in the nginx config above..

Please help all my users are complaining .. :confused:

**Comment by wobbit** [comment]
I'm having the same issue since upgrading to 3.8.1

**Comment by jasonblais** [comment]
@ChrisFernandez @wobbit 

Can you temporarily remove the Site URL field to see if the WebSocket issues stop? This let's us confirm if it's an issue with the Site URL field.

**Comment by wobbit** [comment]
I've *kinda* fixed the issue by removing the port number from the Site URL. Instead I'm now getting a few of these: 

[EROR] websocket routing error: seq=1 uid= api.web_socket_router.not_authenticated.app_error [details: ]

**Comment by wobbit** [comment]
Oh and also I set the AllowCorsFrom to "*" temporarily.

**Comment by hmhealey** [comment]
What is your SiteURL set to? It should match the URL of the proxy (including the http/https protocol and, if necessary, the port) and not the internal address of the Mattermost server. If you have AllowCorsFrom set, it should just be the URL without the protocol or port.

**Comment by ChrisFernandez** [comment]
@wobbit @hmhealey @jasonblais just to make it clear, I did try first the url that is used from the outside and with the https etc only reason my example above is localhost is because when it did not work I try all possible convinations.. on the positive side.. **seems** that if I just leave it empty like it was and **ignore** the warning pressing the "x" everything seems to be working fine.. so maybe you guys did not enforce it yet? in any case this should be check/fix because if I am supposed to add a URL and obviosly only way to be functional is to not add it.. there is an issue :) Also I have two domains pointing to my server... and I see no option in the field to add more than one.. as of now nginx was taking care of that and mattermost did not care since is just listening to localhost.
Thanks for the reply's hopefully we can fix  this between all of us.

**Comment by wobbit** [comment]
@hmhealey it's set the match the URL of the proxy. I tried changing the AllowCorsFrom to the same as the site URL without the protocol or port and still got 

`Apr 21 08:49:10 platform: [2017/04/21 08:49:10 BST] [EROR] websocket connect err: websocket: origin not allowed`

Changing the AllowCorsFrom back to "*" seems to have _fixed_ that problem but we do still get these in the log

`Apr 21 01:54:30 platform: [2017/04/21 01:54:30 BST] [EROR] websocket routing error: seq=1 uid= api.web_socket_router.not_authenticated.app_error [details: ]`

**Comment by hmhealey** [comment]
@ChrisFernandez The Site URL setting is now required due to a security issue arising from the old behaviour so you'll probably want to set it. Not having it set will also cause certain issues like making emails contain broken links. That being said, we've had reports of websocket issues when the Site URL is set so we're currently cutting a 3.8.2 release that will make the websocket always use the URL in the address bar rather than using the Site URL.

In terms of having multiple external URLs, we don't explicitly support that for Mattermost. It may work with 3.8.2, but you won't be able to use the `AllowCorsFrom` setting.

**Comment by hmhealey** [comment]
@wobbit I'm not sure why AllowCorsFrom still wouldn't be working if it matches the Site URL, but for the other issue, can you try logging out and back in again? For whatever reason, the websocket connection isn't linked to a session correctly

**Comment by wobbit** [comment]
@hmhealey It is a bit of an odd one... Tried logging in and out, restarting the desktop app and a fresh browser session yet still getting the same error (but only occasionally). As far as I can see the Mattermost is working fine, except for these errors in the log!

**Comment by jasonblais** [comment]
@ChrisFernandez @wobbit  Mattermost 3.8.2 is available for download here: http://about.mattermost.com/download/

The dot release should fix the WebSocket connection issues following the upgrade. 

Please let us know if you hit any further issues, and thank you for the reports!

**Comment by ohlookemus** [comment]
@jasonblais I upgraded and I'm still seeing this issue.

**Comment by wobbit** [comment]
@jasonblais @ohlookemus  This seems to have fixed it for me when I set my AllowCorsFrom to the same URL as SiteURL. 

If I leave AllowCorsFrom to be ""  it does still error (I'm not sure if this is intended?).

Thanks for the quick fix!

**Comment by jasonblais** [comment]
@ChrisFernandez

Do you have `AllowCorsFrom` setting set for cross-origin requests: https://docs.mattermost.com/administration/config-settings.html#enable-cross-origin-requests-from

If you're using multiple domains, you should add all the alternative domains to AllowCorsFrom setting after which the websockets would work. @wobbit not sure if you're using just a single domain?

Let us know if the issue keeps persisting

**Comment by ChrisFernandez** [comment]
@wobbit @hmhealey @jasonblais  sorry for the late reply, I upgraded to the latest version and that worked just fine for me.. I have not check the logs yet.. but is been working perfectly for now.
@jasonblais I have not set anything on AllowCorsFrom and seems to be working... in any case I will add my 2 domains just in case... I been using TLS with nginx so have not enable TLS on mattermost yet.. such that I want because I am tired of having tor renew TLS manually every 3 months :)

**Comment by ChrisFernandez** [comment]
hmmm I added my two domains and now it broke again... so reverted and left it empty again.

```
[2017/04/24 12:21:30 PDT] [EROR] websocket connect err: websocket: origin not allowed
[2017/04/24 12:21:30 PDT] [EROR] /api/v3/users/websocket:connect code=500 rid=pnryqs4ea**** uid=gohxdcu*** ip=64.71.13.250 Failed to upgrade websocket connection [details: ]
[2017/04/24 12:21:34 PDT] [EROR] websocket connect err: websocket: origin not allowed
[2017/04/24 12:21:34 PDT] [EROR] /api/v3/users/websocket:connect code=500 rid=aefqe1sxzfyzjb69**** uid=gohxdc***  ip=64.71.13.250 Failed to upgrade websocket connection [details: ]
```

**Comment by hmhealey** [comment]
@ChrisFernandez I think AllowCorsFrom only supports a single domain name. I'll double check with the guy who implemented it

**Comment by hmhealey** [comment]
Never mind, that's actually a bug. The [arguments to the function are backwards](https://github.com/mattermost/platform/blob/be9624e2adce7c95039e62fc4ee22538d7fa2d2f/utils/api.go#L18) so it has only ever supported one domain name even though it should allow two. I've filed a ticket to fix that [here](https://mattermost.atlassian.net/browse/PLT-6393)

**Comment by jasonblais** [comment]
@ChrisFernandez curious if you were interested helping with a fix so that AllowCorsFrom supports multiple domains? If not, totally understand.

Ticket here in case you're interested to take a look: https://mattermost.atlassian.net/browse/PLT-6393

**Comment by ohlookemus** [comment]
I had in my address a wildcard like so, *.mydomain.com and I was getting CORs errors. Set it to * and no longer seeing the issue. Has anyone else tried a wildcard domain?

**Comment by ChrisFernandez** [comment]
@jasonblais sure, may take a bit for me to reply depending on day job but of course.

**Comment by jasonblais** [comment]
Awesome! Thanks @ChrisFernandez 👍

**Comment by lindy65** [comment]
Closing this issue off as [ticket](https://mattermost.atlassian.net/browse/PLT-6393) is done and closed and will ship with v3.9 on May 16th.

**Comment by jiyu8iighx** [comment]
> What is your SiteURL set to? It should match the URL of the proxy (including the http/https protocol and, if necessary, the port) and not the internal address of the Mattermost server. If you have AllowCorsFrom set, it should just be the URL without the protocol or port.

Today (ver 8.1.9) I found that I need domain:port, though.
